### PR TITLE
IR 2148 Material Entities for All Meshes, Improved hashing logic, Plugin Components

### PR DESCRIPTION
**Pull request by AidanCaruso**
## Summary
This allows editing and assigning materials using the ECS API to non-gltf sourced meshes, such as the primitive mesh component.

## References
closes [IR-2148](https://tsu.atlassian.net/browse/IR-2148?atlOrigin=eyJpIjoiMzg3NzMxZjdlN2RkNDdlN2ExZDg5ZTU2ODBiZmYwOGYiLCJwIjoiaiJ9), [IR-2390](https://tsu.atlassian.net/browse/IR-2390)

## QA Steps

[IR-2148]: https://tsu.atlassian.net/browse/IR-2148?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[IR-2390]: https://tsu.atlassian.net/browse/IR-2390?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/engine/src/avatar/components/TransparencyDitheringComponent.tsx [hunk 1]
```diff
@@ -76,36 +81,51 @@ export const TransparencyDitheringComponent = Array.from({ length: maxDitherPoin
...
-      shader.fragmentShader = shader.fragmentShader.replace(
-        /#include <common>/,
-        '#include <common>\n' + ditheringFragUniform
+  },
+
+  reactor: () => {
+    const entity = useEntityContext()
+    const materialComponent = getComponent(entity, MaterialComponent[MaterialComponents.State])
+    const material = materialComponent.material as Material
+    addOBCPlugin(material, (shader) => {
```
Comment (HexaField) [inline packages/engine/src/avatar/components/TransparencyDitheringComponent.tsx]: this should be a in a useEffect

Comment (AidanCaruso) [reply packages/engine/src/avatar/components/TransparencyDitheringComponent.tsx]: that is the plan - this is unfinished and i basically just pasted it in a reactor forgetting to put it in a useeffect callback

**Comment by dinomut1** [comment]
Is there a paired PR for converting advanced materials over?

**Comment by AidanCaruso** [comment]
> Is there a paired PR for converting advanced materials over?

yes, it should be linked to this PR already, but here's the PR https://github.com/theinfinitereality/eepro-advanced-materials/pull/33

**Comment by dinomut1** [comment]
When trying to add material plugins i'm not seeing their plugin reactors getting triggered

**Comment by dinomut1** [comment]
[Screencast from 06-14-2024 01:24:03 PM.webm](https://github.com/EtherealEngine/etherealengine/assets/94419856/1197745d-75ea-4782-afd4-f9b7badff9c9)

**Review by dinomut1 — DISMISSED** [review]
I'm seeing a failure in the re-import of a glb with a custom material + plugin set. Here's a video showing my process for reproduction:

[Screencast from 06-17-2024 02:26:25 PM.webm](https://github.com/EtherealEngine/etherealengine/assets/94419856/a3438504-be6e-445e-8072-b63cb8aa91f1)

Relevant stack trace: 
```

QueryFunctions.tsx:192 Uncaught error: TypeError: Cannot read properties of undefined (reading 'length')
    at stringHash (MathFunctions.ts:292:11)
    at hashMaterial (materialSourcingFunctions.ts:195:34)
    at removeMaterial (materialSourcingFunctions.ts:145:16)
    at MaterialLibrarySystem.tsx:109:58
    
```

**Comment by AidanCaruso** [comment]
> I'm seeing a failure in the re-import of a glb with a custom material + plugin set. Here's a video showing my process for reproduction

Fixed, needed some minor adjustments to plugin parameter handling and hashing logic.

**Review by JT00y — COMMENTED** [review]
Screenshot 2024-06-18 12:39:48

I m seeing this error when drag material asset into the scene.

**Comment by AidanCaruso** [comment]
> I m seeing this error when drag material asset into the scene.

This error occurs on dev too and should be addressed in a future PR.

Material drag and drop will give this error if the raycast hits something in the way. In the default scene this error is reproduced by the camera fade transition entity intersecting the ray - this does not have the material instance component, causing the error. Material drag and drop works if there's no intersection.

**Comment by JT00y** [comment]
@AidanCaruso  thanks for explanation for the drag asset. 
Another issue i m seeing was the material library update correct counts of materials only after i save the scene.

https://github.com/EtherealEngine/etherealengine/assets/142644522/7b4d269b-0cd4-403a-8f6f-a71b01b70fcd

**Comment by AidanCaruso** [comment]
> @AidanCaruso thanks for explanation for the drag asset. Another issue i m seeing was the material library update correct counts of materials only after i save the scene.

No problem! This is odd in that I haven't seen it. It looks like one of the meshes is always selected, and the material library is designed to only show the material entities used by the currently selected entity(s) meshes. If you were to press the ESC key do the other material entities still not appear?

**Comment by JT00y** [comment]
> > @AidanCaruso thanks for explanation for the drag asset. Another issue i m seeing was the material library update correct counts of materials only after i save the scene.
> 
> No problem! This is odd in that I haven't seen it. It looks like one of the meshes is always selected, and the material library is designed to only show the material entities used by the currently selected entity(s) meshes. If you were to press the ESC key do the other material entities still not appear?

okay, i understand our design for material library i m thinking maybe the ui part can be better, 
so the problem become reload the scene all the entites will show in library and after click entity it look correct, this probably not a problem from user side.

**Comment by AidanCaruso** [comment]
> okay, i understand our design for material library i m thinking maybe the ui part can be better, so the problem become reload the scene all the entites will show in library and after click entity it look correct, this probably not a problem from user side.

I agree, the UX definitely needs work. First thing we may want to add is a global/local toggle for clarity. For now pressing the escape key to deselect the currently selected entity shows all scene materials again.

**Review by dinomut1 — APPROVED** [review]
works well enough, still some bugs that can be addressed in subsequent PRs

**Pull request merged**